### DISCUSSION TITLE
MVP를 위한 도메인 정의

### DISCUSSION THREAD

**Discussion by Hyune-c** [f-lab-edu/twitch-clone#2] [General]
MVP 범위는 회원입니다.
회원에 관련된 최대한 많은 유스케이스를 댓글로 작성해주시면 됩니다.
기술적인 내용은 가능한 배제하지만, 필요한 경우 기록하셔도 됩니다.

실무에서는 필요한 기능이나 기술적 난이도가 낮거나 학습 필요성이 없다고 생각되는 기능은 축소시키거나 배제할 수 있습니다.
- ex. 회원 삭제

진행 상황을 보고 본문에 업데이트를 하거나, 위키로 이동 시키겠습니다.

https://github.com/f-lab-edu/twitch-clone/wiki/%5B%EC%97%85%EB%AC%B4-%EB%A1%9C%EC%A7%81%5D-%ED%9A%8C%EC%9B%90

**Comment by Hyune-c** [comment]
# 회원

## 공통
1. 회원 종류는 크게 일반/스트리머/구독자/어드민이 있다.
2. 모든 기능은 등록일시와 수정일시가 관리된다.
3. 스트리머 등록은 1회만 가능하다.
4. 구독은 스트리머 등록을 한 회원당 1회만 가능하다.

## 회원 가입 
1. 기 가입 여부 확인이 필요하다.
3. 최소한의 본인 인증이 필요하다. (email)
4. 최소 필요 정보는 email, password, nickName 이다.
    - password는 8자 이상, 문자, 숫자, 특수문자가 각 1개 이상 들어가야 한다.
    - nickName은 12자 이하이다.

## 인증/인가
1. 인증은 지속적인 유지가 필요하다.
3. 인가는 회원 단위로 구분된다.

## 내 정보
1. 상세 정보를 수정할 수 있다. + 비밀번호
    - 분실한 비밀번호는 초기화만 가능하며 이메일로 전달된다.
4. 내가 스트리머라면 내 구독자의 제한된 정보를 조회할 수 있다.
6. 내가 구독자라면 내가 구독한 스트리머의 제한된 정보를 조회할 수 있다.

## 스트리머 
1. 회원명과 구분되는 streamerName을 가진다.
3. streamerName은 고유하다.
4. 구독 비용을 등록할 수 있다.

## 어드민
1. 비밀번호를 제외한 모든 회원 정보를 검색할 수 있다.
2. 비밀번호는 초기화만 가능하다.

## 기타
1. 구독자는 회원 정의 후, 구독에 대한 정의 후 진행한다.

**Comment by DoLiLu** [comment]
### 회원 가입

1. 회원 타입은 일반회원, 스트리머, 어드민이 있다.
2. 사용자는 이메일과 패스워드를 통하여 일반회원으로 가입할 수 있다.
3. 하나의 이메일로 하나의 계정만 회원가입이 가능하다.
4. 회원가입 필수 사항은 이메일, 패스워드, 닉네임이다.
5. 패스워드는 길이 8자이상, 영문, 숫자, 특수문자를 각 1개이상 포함하여야한다.
6. 회원가입 시 입력한 이메일이 존재하는 이메일인지 확인절차가 필요하다.
7. 어드민의 경우 일반 회원가입과 다른 방법으로 회원가입을 진행한다.
8. 어드민의 회원 가입 정책은 일반 사용자와 같다. (이메일, 패스워드)

- 회원 가입시 필수 정보 외에 회원이 갖게 될 부가 정보에 대한 정의 필요.

### 로그인

1. 사용자는 이메일과 패스워드를 통하여 로그인 할 수 있다.
2. 사용자는 로그인 성공/실패 결과를 알 수 있다.
3. 로그인한 정보는 로그아웃 전까지 or 설정된 로그인 만료 기간까지 유지 되어야 한다.
4. 프로그램은 사용자의 로그인 여부를 알 수 있어야 한다.


### 로그인 안 한 사용자

1. 로그인을 하지 않은 사용자는 스트리밍하는 방송을 볼 수 있다.


### 일반 사용자

1. 로그인 성공 시 로그인 한 계정이 획득한 기능을 사용할 수 있다.
2. 사용자는 내 정보를 확인 할 수 있다.
3. 사용자는 내 정보 수정을 할 수 있다. (password, 닉네임)
4. 사용자는 채팅 기능을 사용할 수 있다.
5. 사용자는 스트리밍하는 방송을 볼 수 있다.
6. 사용자는 스트리머를 팔로우 할 수 있다.
7. 사용자는 원하는 스트리머를 구독할 수 있다.
9. 사용자는 구독중인 스트리머를 구독 종료할 수 있다.
10. 스트리머를 하려고 하는 사용자는 스트리머 신청을 할 수 있다.
11. 사용자는 패스워드 초기화 요청을 할 수 있다.


### 스트리머

1. 스트리머는 스트리밍을 할 수 있다.
2. 스트리머는 구독 비용을 결정할 수 있다.
3. 스트리머는 다른 스트리머를 구독 할 수 있다.(기능에 대한 확인이 필요한 부분)
4. 스트리머는 방송 일정을 등록할 수 있다.
5. 스트리머는 추가적인 정보를 등록할 수 있다.(SNS LINK, 스트리머 설명)
6. 스트리머는 팔로워를 가질 수 있으며 총 팔로워 수를 정보로 갖는다.
7.스트리머는 패스워드 초기화 요청을 할 수 있다.

### 어드민
1. 어드민은 스트리머 요청을 승인할 수 있다.
2. 어드민 계정은 스트리머 신청을 할 수 없다.
3. 어드민은 사용자에 대한 정보를 조회 할 수 있다.
4. 어드민은 신고당한 계정을 강제 탈퇴 시킬 수 있다.
5. 어드민은 신고당한 스트리머의 스트리머 권한을 정지 시킬 수 있다.
6. 어드민은 요청받은 사용자의 패스워드를 초기화 하고 초기화 한 패스워드를 사용자에게 전달할 수 있다.

**Comment by FrancescoJo** [comment]
회원은 어떤 속성(Property) 들을 가지는게 좋을까요?
